src/Workspaces/Core/Portable/Workspace/Solution/Checksum_Factory.cs

```diff
@@ -23,47 +22,102 @@ internal partial class Checksum
+#else
+        private static readonly ObjectPool<SHA256> s_incrementalHashPool =
+            new(SHA256.Create, size: 20);
+#endif
 
+#if !NET5_0_OR_GREATER
         // Dedicated pools for the byte[]s we use to create checksums from two or three existing checksums. Sized to
         // exactly the space needed to splat the existing checksum data into the array and then hash it.
 
         private static readonly ObjectPool<byte[]> s_twoChecksumByteArrayPool = new(() => new byte[HashSize * 2]);
         private static readonly ObjectPool<byte[]> s_threeChecksumByteArrayPool = new(() => new byte[HashSize * 3]);
+#endif
 
         public static Checksum Create(IEnumerable<string> values)
         {
+#if NET5_0_OR_GREATER
+            using var pooledHash = s_incrementalHashPool.GetPooledObject();
+
+            foreach (var value in values)
+            {
+                pooledHash.Object.AppendData(MemoryMarshal.AsBytes(value.AsSpan()));
+                pooledHash.Object.AppendData(MemoryMarshal.AsBytes("\0".AsSpan()));
+            }
+
+            Span<byte> hash = stackalloc byte[SHA256HashSizeBytes];
+            pooledHash.Object.GetHashAndReset(hash);
+            return From(hash);
+#else
             using var pooledHash = s_incrementalHashPool.GetPooledObject();
             using var pooledBuffer = SharedPools.ByteArray.GetPooledObject();
             var hash = pooledHash.Object;
 
+            hash.Initialize();
             foreach (var value in values)
             {
                 AppendData(hash, pooledBuffer.Object, value);
                 AppendData(hash, pooledBuffer.Object, "\0");
             }
 
-            return From(hash.GetHashAndReset());
+            hash.TransformFinalBlock(Array.Empty<byte>(), 0, 0);
+            return From(hash.Hash);
+#endif
         }
 
         public static Checksum Create(string value)
         {
+#if NET5_0_OR_GREATER
+            Span<byte> hash = stackalloc byte[SHA256HashSizeBytes];
+            SHA256.HashData(MemoryMarshal.AsBytes(value.AsSpan()), hash);
+            return From(hash);
+#else
             using var pooledHash = s_incrementalHashPool.GetPooledObject();
             using var pooledBuffer = SharedPools.ByteArray.GetPooledObject();
             var hash = pooledHash.Object;
+            hash.Initialize();
 
             AppendData(hash, pooledBuffer.Object, value);
 
-            return From(hash.GetHashAndReset());
+            hash.TransformFinalBlock(Array.Empty<byte>(), 0, 0);
+            return From(hash.Hash);
+#endif
         }
 
         public static Checksum Create(Stream stream)
         {
+#if NET7_0_OR_GREATER
+            Span<byte> hash = stackalloc byte[SHA256HashSizeBytes];
+            SHA256.HashData(stream, hash);
+            return From(hash);
+#elif NET5_0_OR_GREATER
+            using var pooledHash = s_incrementalHashPool.GetPooledObject();
+            Span<byte> buffer = stackalloc byte[SharedPools.ByteBufferSize];
```

Comment:
4K is a much larger buffer than we typically like to stackalloc.  It might be fine for your scenarios, just calling it out.  We have a rough guideline of not stackalloc'ing more than 1K.

Reply:
(Also, keep in mind this whole buffer will be zero'd unless you're using SkipLocalsInit.)